### ISSUE TITLE
Flickering when moving the mouse in DX12/Win32

### ISSUE THREAD

**Issue by b1skit** [ocornut/imgui#7238]
### Version/Branch of Dear ImGui:

1.89 master

### Back-ends:

imgui_impl_dx12.h + imgui_impl_win32.h

### Compiler, OS:

Windows 10 + MSVC 2022

### Full config/build information:

```
Dear ImGui 1.89.8 (18980)
--------------------------------
sizeof(size_t): 8, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=199711
define: _WIN32
define: _WIN64
define: _MSC_VER=1937
define: _MSVC_LANG=202002
--------------------------------
io.BackendPlatformName: imgui_impl_win32
io.BackendRendererName: imgui_impl_dx12
io.ConfigFlags: 0x00000000
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x0000000E
 HasMouseCursors
 HasSetMousePos
 RendererHasVtxOffset
--------------------------------
io.Fonts: 1 fonts, Flags: 0x00000000, TexSize: 512,64
io.DisplaySize: 1920.00,1080.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 8.00,8.00
style.WindowBorderSize: 1.00
style.FramePadding: 4.00,3.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 8.00,4.00
style.ItemInnerSpacing: 4.00,4.00
```

### Details:

I'm getting flickering when moving the mouse in Release builds of my custom rendering engine. The issue seems kind of similar to this one https://github.com/ocornut/imgui/issues/4711, although that could be coincidental and I'm using the standard ImGui backend implementations with DX12/Win32 so I'm creating a new ticket.

I'm not doing anything fancy or clever with ImGui. My engine is multi-threaded, but I use a command queue of function lambdas to ensure all ImGui calls are deferred until they can be executed on the render thread. I use a single global mutex to wrap all accesses to ImGui - on the render thread, or otherwise.

It feels very much like a race condition on the CPU side (especially since it does not occur in debug builds), but I don't see how that is possible. I've carefully reviewed all of my ImGui code (there's not much) and all ImGui accesses (e.g. to set keyboard or mouse states) are wrapped in my global mutex.

The render thread locks the same mutex, and holds it the entire time from the ImGui NewFrame call, until ImGui's command list is executed.

I've tried forcing serial execution of ImGui rendering calls by waiting on a fence for ImGui to complete before progressing further, but this does not help.

The flickering seems directly related to mouse motion. If I don't move the mouse, no flickering occurs. Interestingly, the io.WantCaptureMouse variable seems to flicker on/off in time with the flickering, I've attached a .gif of the behavior. This makes it hard to select things/keep them selected.

I'm unable to reproduce the behavior in the DX12/Win32 sample, but I've compared that with my implementation and I believe they're functionally identical (I copy pasted it when doing my original implementation).

I've taken renderdoc captures of this issue, and when the flickering occurs it appears as if ImGui is not being drawn at all:
![image](https://github.com/ocornut/imgui/assets/8851500/17aff7ed-f3eb-427b-ab4f-0922336449da)
However, I've set debug asserts and verified that I'm executing the same ImGui render code every single frame without fail when the flickering is happening. 

The RenderDoc capture shows the command list being executed, but no draws are recorded. Here are 2 consecutive frames when the flicker occurs:
![image](https://github.com/ocornut/imgui/assets/8851500/0f01b267-828b-4e21-8fa9-1b645ea3f054)
![image](https://github.com/ocornut/imgui/assets/8851500/712a91b3-18ba-49d0-a061-4c5c2a55ba0c)

This is a personal project, so I'm happy to share snippets of code if needed.

Is there anything I can do to try an narrow down what's happening here? 

### Screenshots/Video:

![bug](https://github.com/ocornut/imgui/assets/8851500/84e72614-1ee7-46c7-8944-7fa8dea1fb0f)

### Minimal, Complete and Verifiable Example code:

```cpp
// Here's some code anyone can copy and paste to reproduce your issue
ImGui::Begin("Example Bug");
MoreCodeToExplainMyIssue();
ImGui::End();
```

**Comment by GamingMinds-DanielC** [comment]
Hard to say without any code, but I'm pretty sure the problem must be in your synchronization. A "command queue of function lambdas" implies that your lambdas will be dismissed once executed, it wouldn't be a queue otherwise. Thread scheduling is very complex and there is no guarantee regarding execution order and distribution of processing time, any thread can be interrupted at any time. A mutex only guarantees mutually exclusive access to resources. Very dependent on timing and therefore hard to debug.

What most likely happens here is something like this: your render thread, after rendering a frame and releasing the mutex, reacquires the mutex to render the next frame before any of your other threads got in to queue up a function lambda. Try analyzing your queue every frame. The queue size alone might not be enough, one thread failing to queue up a lambda could be offset by another submitting more than once.

**Comment by ocornut** [comment]
> io.WantCaptureMouse variable seems to flicker on/off in time with the flickering, 

That doesn't depend at all on backends, and is a strong hint that the Dear ImGui state is changing, aka maybe you are submitting empty frames aka calling `ImGui::NewFrame()` with no contents.

Try to add some logging about your `ImGui::EndFrame()` or `ImGui::Render()` calls:

```cpp
printf("MetricsActiveWindows=%d\n", ImGuI::GetIO().MetricsActiveWindows);
```

**Comment by b1skit** [comment]
Thank you @GamingMinds-DanielC and @ocornut  - indeed, this was a synchronization issue in my app and not an ImGui issue. Both of your suggestions helped me narrow it down, it was trivial to fix once I spotted the issue :)
